internal/cli/cli.go

```diff
@@ -0,0 +1,88 @@
+package cli
+
+import (
+	"io"
+
+	"github.com/10gen/realm-cli/internal/config"
+	"github.com/10gen/realm-cli/internal/realm"
+	"github.com/10gen/realm-cli/internal/ui"
+
+	"github.com/spf13/cobra"
+)
+
+// Command is an executable CLI command
+type Command interface {
+	Handler(args []string, io cmdio) error
+}
+
+// CommandPreparer handles the command setup phase
+type CommandPreparer interface {
+	Setup(args []string, io cmdio) error
+}
+
+// CommandResponder handles the command feedback phase
+type CommandResponder interface {
+	Feedback(args []string, io cmdio)
+}
+
+type cmdio struct {
+	in  io.Reader
+	err io.Writer
+	out io.Writer
+}
+
+func newCmdio(cmd *cobra.Command) cmdio {
+	return cmdio{
+		out: cmd.OutOrStdout(),
+		err: cmd.ErrOrStderr(),
+		in:  cmd.InOrStdin(),
+	}
+}
+
+func newCommand(use, short string, cmd Command) *cobra.Command {
+	var preRun cmdFnE
+	if c, ok := cmd.(CommandPreparer); ok {
+		preRun = c.Setup
+	}
+
+	var postRun cmdFn
+	if c, ok := cmd.(CommandResponder); ok {
+		postRun = c.Feedback
+	}
+
+	return &cobra.Command{
+		Use:     use,
+		Short:   short,
+		PreRunE: cobraFnE(preRun),
+		RunE:    cobraFnE(cmd.Handler),
+		PostRun: cobraFn(postRun),
```

Comment:
So I went this route to intentionally limit the Feedback phase to be "error-free" (I don't want to report a "failed command" just because printing results failed, the server-side-effects likely still occurred).

I feel like that's something still worth preserving, over the benefit of simplicity/consistency here.  Thoughts?

Reply:
Edit: using the `PostRunE` hook, and proper error handling for the feedback stage will be addressed in [REALMC-7340](https://jira.mongodb.org/browse/REALMC-7340)